pose_evaluation/metrics/signclip_distance_metric.py

```diff
@@ -0,0 +1,49 @@
+from pose_evaluation.metrics.base_embedding_metric import NumpyArrayEmbeddingMetric
+from typing import Literal
+import numpy as np
+from tqdm import tqdm
+from scipy.spatial.distance import cosine
+import torch
+import torch.nn.functional as F
+
+
+class SignCLIPEmbeddingDistanceMetric(NumpyArrayEmbeddingMetric):
+    def __init__(self, kind: str = "cosine", device: torch.device | str = "cuda"):
+        """
+        Initializes the metric with the specified distance type and device.
+
+        Args:
+            kind (str): The type of distance metric, either 'cosine' or 'l2'.
+            device (torch.device | str): The device to use ('cuda' or 'cpu').
+        """
+        self.kind = kind
+        self.device = torch.device(device) if isinstance(device, str) else device
+
+    def score_all(self, embeddings: torch.Tensor) -> torch.Tensor:
+        """
+        Computes the pairwise distance matrix for the provided embeddings.
+
+        Args:
+            embeddings (torch.Tensor): A 2D tensor of shape (N, D), where N is the number
+                                        of embeddings and D is the feature dimension.
+
+        Returns:
+            torch.Tensor: A 2D tensor of shape (N, N) containing pairwise distances.
+        """
+        # Move embeddings to the specified device
+        embeddings = embeddings.to(self.device)
+
+        if self.kind == "cosine":
+            # Normalize embeddings to unit norm
+            embeddings = F.normalize(embeddings, p=2, dim=1)
+            # Compute pairwise cosine similarity
+            similarity_matrix = torch.matmul(embeddings, embeddings.T)  # Shape: (N, N)
+            distance_matrix = 1 - similarity_matrix  # Cosine distance = 1 - cosine similarity
```

Comment:
use cosine similarity https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html

Reply:
Did some benchmarking, this way (with matmul) gives the same results and is faster... At least on CPU. On CUDA the result is apparently almost the same either way.

https://colab.research.google.com/drive/1wMFIrxdXwOS5p1cuot0-JLXF9zmCUpVB?usp=sharing

Reply:
Given that we need to use GPU anyway and this is nicer, I think I will change to the F.cosine_similarity

Reply:
I have to mention that during the training and inference of both the original VideoCLIP and SignCLIP, we have always been using dot product instead of cosine similarity. A while ago we discussed it but at the end for SignCLIP dot product just works better empirically. We might want to revisit this choice and understand better about it.

At the moment, it would be good to also support dot product here, just skip the two `F.normalize` calls before `torch.matmul`. We can also again test their differences empirically.

Reply:
So, @J22Melody if I understand correctly is proposing three types of similarity:
1. Euclidean
2. dot
3. cosine

my only concern with "dot" and euclidean is that both are not normalized, and that makes the "distance" unbounded

Reply:
yes, it is an issue, it is possible to do a re-normalization on the scores afterward though. But if cosine similarity turns out to work well, we can just keep it as default as it gives -1 to 1 scores.

Any other ideas?

Reply:
Problem with bounded vs unbounded:
- When performing top k, it is not a problem. The "top k" is a 0/1 metric per sample.
- When performing similarity, it is a problem: The similarity between every two samples is unbounded, and so when averaging, you get a skewed number

Reply:
Looking at it from comparability/compatibility:
* pgvector uses distances, not similarities. 
* I think a lot of the classic works on this do use cosine distance, including the OG "king queen man woman" one: http://arxiv.org/abs/1301.3781.

On the other hand there are some that use cosine similarity too, like
* https://arxiv.org/pdf/1702.01417 

Looking at it from popularity: 
* Semantic Scholar has about 2 million results for "cosine distance" and about a million for "cosine similarity". 

Looking at it for our use case:
* We want to find the distribution of scores by finding nearest neighbors, right? (https://github.com/sign-language-processing/pose-evaluation?tab=readme-ov-file#distribution-of-scores)
* So would it be the case that either one works? unbounded distances are fine as mentioned, since we're effectively doing top K. 
* However it does mean that we can't really compare numbers between metrics, I guess? If Euclidean has a distance of "one million" and Cosine has a distance of 0.9, it is harder to compare them, though like Zifan said you can renormalize afterwards. 

I guess I'll go ahead with distances for now

Reply:
Whatever you do - we can implement a cosine similarity metric as `1 - ` so it is just your choice. Distance metric is more generic as it can also do Euclidean and dot

Reply:
For now, we might want to use dot product because it works the best empirically. And yes you can do standardization (minus mean and divide by std) or min-max normalization to re-scale the numbers to a certain range.

What I tried during the SignCLIP experiments:

- add a multimodal projection layer then L2 norm (as done for CLIP), the training was terrible 
- add a multimodal projection layer only, the training is okay, but also does seem to be beneficial (E5.1)

I still do not understand why and I did not try training with the L2 norm on the pooled embeddings without the multimodal projection layer. If this could work, then we will have the nice properties of unit norm embeddings:

- cosine similarity equals dot product
- cosine similarity is also essentially equivalent to the Euclidean distance (see the screenshot below)

<img width="770" alt="image" src="https://github.com/user-attachments/assets/22664c21-2639-4634-af58-347d29b39204">